CircuitBreakerServiceIT.testParentChecking failures

https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.1+multijob-unix-compatibility/os=opensuse/191/consoleFull

The failure is not reproducible but seems to only affect 6.1.

```
12:35:56   1> org.elasticsearch.transport.RemoteTransportException: [node_td3][127.0.0.1:39190][indices:data/read/search[phase/query/id]]
12:35:56   1> Caused by: org.elasticsearch.search.query.QueryPhaseExecutionException: Query Failed [Failed to execute main query]
12:35:56   1> 	at org.elasticsearch.search.query.QueryPhase.execute(QueryPhase.java:298) ~[main/:?]
12:35:56   1> 	at org.elasticsearch.search.query.QueryPhase.execute(QueryPhase.java:110) ~[main/:?]
12:35:56   1> 	at org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:420) ~[main/:?]
12:35:56   1> 	at org.elasticsearch.action.search.SearchTransportService$7.messageReceived(SearchTransportService.java:404) ~[main/:?]
12:35:56   1> 	at org.elasticsearch.action.search.SearchTransportService$7.messageReceived(SearchTransportService.java:401) ~[main/:?]
12:35:56   1> 	at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:66) ~[main/:?]
12:35:56   1> 	at org.elasticsearch.transport.TcpTransport$RequestHandler.doRun(TcpTransport.java:1554) ~[main/:?]
12:35:56   1> 	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:637) ~[main/:?]
12:35:56   1> 	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) ~[main/:?]
12:35:56   1> 	at org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:41) ~[main/:?]
12:35:56   1> 	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) ~[main/:?]
12:35:56   1> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[?:1.8.0_151]
12:35:56   1> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[?:1.8.0_151]
12:35:56   1> 	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_151]
12:35:56   1> Caused by: org.elasticsearch.ElasticsearchException: java.util.concurrent.ExecutionException: CircuitBreakingException[[fielddata] Data too large, data for [test] would be [1758/1.7kb], which is larger than the limit of [10/10b]]
12:35:56   1> 	at org.elasticsearch.index.fielddata.plain.AbstractIndexFieldData.load(AbstractIndexFieldData.java:74) ~[main/:?]
12:35:56   1> 	at org.elasticsearch.index.fielddata.fieldcomparator.BytesRefFieldComparatorSource$1.getSortedDocValues(BytesRefFieldComparatorSource.java:87) ~[main/:?]
12:35:56   1> 	at org.apache.lucene.search.FieldComparator$TermOrdValComparator.getLeafComparator(FieldComparator.java:714) ~[lucene-core-7.1.0.jar:7.1.0 84c90ad2c0218156c840e19a64d72b8a38550659 - ubuntu - 2017-10-13 16:12:42]
12:35:56   1> 	at org.apache.lucene.search.FieldValueHitQueue.getComparators(FieldValueHitQueue.java:180) ~[lucene-core-7.1.0.jar:7.1.0 84c90ad2c0218156c840e19a64d72b8a38550659 - ubuntu - 2017-10-13 16:12:42]
12:35:56   1> 	at org.apache.lucene.search.TopFieldCollector$SimpleFieldCollector.getLeafCollector(TopFieldCollector.java:100) ~[lucene-core-7.1.0.jar:7.1.0 84c90ad2c0218156c840e19a64d72b8a38550659 - ubuntu - 2017-10-13 16:12:42]
12:35:56   1> 	at org.apache.lucene.search.FilterCollector.getLeafCollector(FilterCollector.java:40) ~[lucene-core-7.1.0.jar:7.1.0 84c90ad2c0218156c840e19a64d72b8a38550659 - ubuntu - 2017-10-13 16:12:42]
12:35:56   1> 	at org.elasticsearch.search.query.CancellableCollector.getLeafCollector(CancellableCollector.java:51) ~[main/:?]
12:35:56   1> 	at org.apache.lucene.search.IndexSearcher.search(IndexSearcher.java:649) ~[lucene-core-7.1.0.jar:7.1.0 84c90ad2c0218156c840e19a64d72b8a38550659 - ubuntu - 2017-10-13 16:12:42]
12:35:56   1> 	at org.elasticsearch.search.internal.ContextIndexSearcher.search(ContextIndexSearcher.java:186) ~[main/:?]
12:35:56   1> 	at org.apache.lucene.search.IndexSearcher.search(IndexSearcher.java:462) ~[lucene-core-7.1.0.jar:7.1.0 84c90ad2c0218156c840e19a64d72b8a38550659 - ubuntu - 2017-10-13 16:12:42]
12:35:56   1> 	at org.elasticsearch.search.query.QueryPhase.execute(QueryPhase.java:272) ~[main/:?]
12:35:56   1> 	at org.elasticsearch.search.query.QueryPhase.execute(QueryPhase.java:110) ~[main/:?]
12:35:56   1> 	at org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:420) ~[main/:?]
12:35:56   1> 	at org.elasticsearch.action.search.SearchTransportService$7.messageReceived(SearchTransportService.java:404) ~[main/:?]
12:35:56   1> 	at org.elasticsearch.action.search.SearchTransportService$7.messageReceived(SearchTransportService.java:401) ~[main/:?]
12:35:56   1> 	at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:66) ~[main/:?]
12:35:56   1> 	at org.elasticsearch.transport.TcpTransport$RequestHandler.doRun(TcpTransport.java:1554) ~[main/:?]
12:35:56   1> 	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:637) ~[main/:?]
12:35:56   1> 	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) ~[main/:?]
12:35:56   1> 	at org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:41) ~[main/:?]
12:35:56   1> 	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) ~[main/:?]
12:35:56   1> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[?:1.8.0_151]
12:35:56   1> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[?:1.8.0_151]
12:35:56   1> 	at java.lang.Thread.run(Thread.java:748) ~[?:1.8.0_151]

12:36:01   2> REPRODUCE WITH: ./gradlew :core:integTest -Dtests.seed=32BD17E618B23F85 -Dtests.class=org.elasticsearch.indices.memory.breaker.CircuitBreakerServiceIT -Dtests.method="testParentChecking" -Dtests.security.manager=true -Dtests.locale=it-CH -Dtests.timezone=Asia/Beirut
```

I looked into this, the failure isn't that a circuit breaker is being tripped (that's expected), it's that there are non-closed contexts:

```
FAILURE 21.1s J2 | CircuitBreakerServiceIT.testParentChecking <<< FAILURES!
   > Throwable #1: java.lang.AssertionError: There are still [2] in-flight contexts. The first one's creation site is listed as the cause of this exception.
   > 	at org.elasticsearch.search.MockSearchService.assertNoInFlightContext(MockSearchService.java:51)
   > 	at org.elasticsearch.test.ESTestCase.lambda$ensureAllSearchContextsReleased$4(ESTestCase.java:419)
   > 	at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:726)
   > 	at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:700)
   > 	at org.elasticsearch.test.ESTestCase.ensureAllSearchContextsReleased(ESTestCase.java:419)
   > 	at org.elasticsearch.test.ESIntegTestCase.cleanUpCluster(ESIntegTestCase.java:2068)
   > 	at java.lang.Thread.run(Thread.java:748)
   > 	Suppressed: java.lang.AssertionError: There are still [2] in-flight contexts. The first one's creation site is listed as the cause of this exception.
   > 		at org.elasticsearch.search.MockSearchService.assertNoInFlightContext(MockSearchService.java:51)
   > 		at org.elasticsearch.test.ESTestCase.lambda$ensureAllSearchContextsReleased$4(ESTestCase.java:419)
   > 		at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:714)
   > 		... 37 more
   > 	Caused by: java.lang.RuntimeException: [dZ_96OWrR4qirRgFGrID6w][cb-test][6]searchType=[DFS_QUERY_THEN_FETCH] query=[*:*]
   > 		at org.elasticsearch.search.MockSearchService.addActiveContext(MockSearchService.java:59)
   > 		at org.elasticsearch.search.MockSearchService.putContext(MockSearchService.java:78)
   > 		at org.elasticsearch.search.SearchService.createAndPutContext(SearchService.java:531)
   > 		at org.elasticsearch.search.SearchService.executeDfsPhase(SearchService.java:275)
   > 		at org.elasticsearch.search.SearchService.access$000(SearchService.java:113)
   > 		at org.elasticsearch.search.SearchService$1.onResponse(SearchService.java:261)
   > 		at org.elasticsearch.search.SearchService$1.onResponse(SearchService.java:257)
   > 		at org.elasticsearch.search.SearchService$3.doRun(SearchService.java:996)
   > 		at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:637)
   > 		at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
   > 		at org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:41)
   > 		at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
   > 		at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
   > 		at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
   > 		... 1 more
   > 	Suppressed: java.lang.AssertionError: There are still [2] in-flight contexts. The first one's creation site is listed as the cause of this exception.
   > 		at org.elasticsearch.search.MockSearchService.assertNoInFlightContext(MockSearchService.java:51)
   > 		at org.elasticsearch.test.ESTestCase.lambda$ensureAllSearchContextsReleased$4(ESTestCase.java:419)
   > 		at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:714)
   > 		... 37 more
```

--- a/server/src/test/java/org/elasticsearch/common/breaker/MemoryCircuitBreakerTests.java
+++ b/server/src/test/java/org/elasticsearch/common/breaker/MemoryCircuitBreakerTests.java
@@ -21,6 +21,7 @@ package org.elasticsearch.common.breaker;
 
 import org.elasticsearch.common.settings.ClusterSettings;
 import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.ByteSizeUnit;
 import org.elasticsearch.common.unit.ByteSizeValue;
 import org.elasticsearch.indices.breaker.BreakerSettings;
 import org.elasticsearch.indices.breaker.CircuitBreakerService;
@@ -31,6 +32,7 @@ import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicReference;
 
+import static org.hamcrest.Matchers.containsString;
 import static org.hamcrest.Matchers.equalTo;
 import static org.hamcrest.Matchers.greaterThanOrEqualTo;
 
@@ -239,6 +241,42 @@ public class MemoryCircuitBreakerTests extends ESTestCase {
             long newUsed = (long)(breaker.getUsed() * breaker.getOverhead());
             assertThat(cbe.getMessage().contains("would be [" + newUsed + "/"), equalTo(true));
             assertThat(cbe.getMessage().contains("field [" + field + "]"), equalTo(true));
+        }
+    }
+
+    /**
+     * Test that a breaker correctly redistributes to a different breaker, in
+     * this case, the request breaker borrows space from the fielddata breaker
+     */
+    public void testBorrowingSiblingBreakerMemory() throws Exception {
+        Settings clusterSettings = Settings.builder()
+                .put(HierarchyCircuitBreakerService.TOTAL_CIRCUIT_BREAKER_LIMIT_SETTING.getKey(), "200mb")
+                .put(HierarchyCircuitBreakerService.REQUEST_CIRCUIT_BREAKER_LIMIT_SETTING.getKey(), "150mb")
+                .put(HierarchyCircuitBreakerService.FIELDDATA_CIRCUIT_BREAKER_LIMIT_SETTING.getKey(), "150mb")
+                .build();
+        try (CircuitBreakerService service = new HierarchyCircuitBreakerService(clusterSettings,
+                new ClusterSettings(clusterSettings, ClusterSettings.BUILT_IN_CLUSTER_SETTINGS))) {
+            CircuitBreaker requestCircuitBreaker = service.getBreaker(MemoryCircuitBreaker.REQUEST);
+            CircuitBreaker fieldDataCircuitBreaker = service.getBreaker(MemoryCircuitBreaker.FIELDDATA);
+
+            assertEquals(new ByteSizeValue(200, ByteSizeUnit.MB).getBytes(),
+                    service.stats().getStats(MemoryCircuitBreaker.PARENT).getLimit());
+            assertEquals(new ByteSizeValue(150, ByteSizeUnit.MB).getBytes(), requestCircuitBreaker.getLimit());
+            assertEquals(new ByteSizeValue(150, ByteSizeUnit.MB).getBytes(), fieldDataCircuitBreaker.getLimit());
+
+            double fieldDataUsedBytes = fieldDataCircuitBreaker
+                    .addEstimateBytesAndMaybeBreak(new ByteSizeValue(50, ByteSizeUnit.MB).getBytes(), "should not break");
+            assertEquals(new ByteSizeValue(50, ByteSizeUnit.MB).getBytes(), fieldDataUsedBytes, 0.0);
+            double requestUsedBytes = requestCircuitBreaker.addEstimateBytesAndMaybeBreak(new ByteSizeValue(50, ByteSizeUnit.MB).getBytes(),
+                    "should not break");
+            assertEquals(new ByteSizeValue(50, ByteSizeUnit.MB).getBytes(), requestUsedBytes, 0.0);
+            requestUsedBytes = requestCircuitBreaker.addEstimateBytesAndMaybeBreak(new ByteSizeValue(50, ByteSizeUnit.MB).getBytes(),
+                    "should not break");
+            assertEquals(new ByteSizeValue(100, ByteSizeUnit.MB).getBytes(), requestUsedBytes, 0.0);
+            CircuitBreakingException exception = expectThrows(CircuitBreakingException.class, () -> requestCircuitBreaker
+                    .addEstimateBytesAndMaybeBreak(new ByteSizeValue(50, ByteSizeUnit.MB).getBytes(), "should break"));
+            assertThat(exception.getMessage(), containsString("[parent] Data too large, data for [should break] would be"));
+            assertThat(exception.getMessage(), containsString("which is larger than the limit of [209715200/200mb]"));
         }
     }
 }

--- a/server/src/test/java/org/elasticsearch/indices/memory/breaker/CircuitBreakerServiceIT.java
+++ b/server/src/test/java/org/elasticsearch/indices/memory/breaker/CircuitBreakerServiceIT.java
@@ -29,7 +29,6 @@ import org.elasticsearch.action.index.IndexRequest;
 import org.elasticsearch.action.index.IndexRequestBuilder;
 import org.elasticsearch.action.search.SearchRequestBuilder;
 import org.elasticsearch.action.search.SearchResponse;
-import org.elasticsearch.action.search.ShardSearchFailure;
 import org.elasticsearch.client.Client;
 import org.elasticsearch.client.Requests;
 import org.elasticsearch.cluster.metadata.IndexMetaData;
@@ -50,7 +49,6 @@ import org.elasticsearch.rest.RestStatus;
 import org.elasticsearch.search.sort.SortOrder;
 import org.elasticsearch.test.ESIntegTestCase;
 import org.elasticsearch.test.ESIntegTestCase.ClusterScope;
-import org.elasticsearch.test.junit.annotations.TestLogging;
 import org.junit.After;
 import org.junit.Before;
 
@@ -73,7 +71,6 @@ import static org.hamcrest.CoreMatchers.instanceOf;
 import static org.hamcrest.Matchers.equalTo;
 import static org.hamcrest.Matchers.greaterThanOrEqualTo;
 import static org.hamcrest.Matchers.nullValue;
-import static org.hamcrest.Matchers.startsWith;
 
 /**
  * Integration tests for InternalCircuitBreakerService
@@ -220,88 +217,6 @@ public class CircuitBreakerServiceIT extends ESIntegTestCase {
             breaks += breakerStats.getTrippedCount();
         }
         assertThat(breaks, greaterThanOrEqualTo(1));
-    }
-
-    /**
-     * Test that a breaker correctly redistributes to a different breaker, in
-     * this case, the fielddata breaker borrows space from the request breaker
-     */
-    @TestLogging("_root:DEBUG,org.elasticsearch.action.search:TRACE")
-    public void testParentChecking() throws Exception {
-        if (noopBreakerUsed()) {
-            logger.info("--> noop breakers used, skipping test");
-            return;
-        }
-        assertAcked(prepareCreate("cb-test", 1, Settings.builder().put(SETTING_NUMBER_OF_REPLICAS, between(0, 1)))
-                .addMapping("type", "test", "type=text,fielddata=true"));
-        Client client = client();
-
-        // index some different terms so we have some field data for loading
-        int docCount = scaledRandomIntBetween(300, 1000);
-        List<IndexRequestBuilder> reqs = new ArrayList<>();
-        for (long id = 0; id < docCount; id++) {
-            reqs.add(client.prepareIndex("cb-test", "type", Long.toString(id)).setSource("test", "value" + id));
-        }
-        indexRandom(true, reqs);
-
-        Settings resetSettings = Settings.builder()
-                .put(HierarchyCircuitBreakerService.FIELDDATA_CIRCUIT_BREAKER_LIMIT_SETTING.getKey(), "10b")
-                .put(HierarchyCircuitBreakerService.FIELDDATA_CIRCUIT_BREAKER_OVERHEAD_SETTING.getKey(), 1.0)
-                .build();
-        assertAcked(client.admin().cluster().prepareUpdateSettings().setTransientSettings(resetSettings));
-
-        // Perform a search to load field data for the "test" field
-        try {
-            client.prepareSearch("cb-test").setQuery(matchAllQuery()).addSort("test", SortOrder.DESC).get();
-            fail("should have thrown an exception");
-        } catch (Exception e) {
-            String errMsg = "CircuitBreakingException[[fielddata] Data too large, data for [test] would be";
-            assertThat("Exception: [" + e.toString() + "] should contain a CircuitBreakingException", e.toString(), containsString(errMsg));
-            errMsg = "which is larger than the limit of [10/10b]]";
-            assertThat("Exception: [" + e.toString() + "] should contain a CircuitBreakingException", e.toString(), containsString(errMsg));
-        }
-
-        // execute a search that loads field data (sorting on the "test" field)
-        // again, this time it should trip the breaker
-        SearchRequestBuilder searchRequest = client.prepareSearch("cb-test").setQuery(matchAllQuery()).addSort("test", SortOrder.DESC);
-
-        String errMsg = "Data too large, data for [test] would be";
-        assertFailures(searchRequest, RestStatus.INTERNAL_SERVER_ERROR, containsString(errMsg));
-        errMsg = "which is larger than the limit of [10/10b]";
-        assertFailures(searchRequest, RestStatus.INTERNAL_SERVER_ERROR, containsString(errMsg));
-
-        reset();
-
-        // Adjust settings so the parent breaker will fail, but neither the fielddata breaker nor the node request breaker will fail
-        resetSettings = Settings.builder()
-                .put(HierarchyCircuitBreakerService.TOTAL_CIRCUIT_BREAKER_LIMIT_SETTING.getKey(), "500b")
-                .put(HierarchyCircuitBreakerService.FIELDDATA_CIRCUIT_BREAKER_LIMIT_SETTING.getKey(), "90%")
-                .put(HierarchyCircuitBreakerService.FIELDDATA_CIRCUIT_BREAKER_OVERHEAD_SETTING.getKey(), 1.0)
-                .build();
-        client.admin().cluster().prepareUpdateSettings().setTransientSettings(resetSettings).execute().actionGet();
-
-        // Perform a search to load field data for the "test" field
-        try {
-            SearchResponse searchResponse = client.prepareSearch("cb-test").setQuery(matchAllQuery()).addSort("test", SortOrder.DESC).get();
-            if (searchResponse.getShardFailures().length > 0) {
-                // each shard must have failed with CircuitBreakingException
-                for (ShardSearchFailure shardSearchFailure : searchResponse.getShardFailures()) {
-                    Throwable cause = ExceptionsHelper.unwrap(shardSearchFailure.getCause(), CircuitBreakingException.class);
-                    assertThat(cause, instanceOf(CircuitBreakingException.class));
-                    assertEquals(((CircuitBreakingException) cause).getByteLimit(), 500L);
-                }
-            } else {
-                fail("should have thrown a CircuitBreakingException");
-            }
-        } catch (Exception e) {
-            Throwable cause = ExceptionsHelper.unwrap(e, CircuitBreakingException.class);
-            assertThat(cause, instanceOf(CircuitBreakingException.class));
-            assertEquals(((CircuitBreakingException) cause).getByteLimit(), 500L);
-            assertThat("Exception: [" + cause.toString() + "] should be caused by the parent circuit breaker",
-                cause.toString(), startsWith("CircuitBreakingException[[parent] Data too large"));
-        }
-
-        reset();
     }
 
     public void testRequestBreaker() throws Exception {

